src/Middleware/RequestDecompression/src/RequestDecompressionProvider.cs

```diff
@@ -0,0 +1,102 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Linq;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.RequestDecompression;
+
+/// <inheritdoc />
+internal sealed class RequestDecompressionProvider : IRequestDecompressionProvider
+{
+    private readonly IDecompressionProvider[] _providers;
+    private readonly ILogger _logger;
+
+    /// <summary>
+    /// If no decompression providers are specified then all default providers will be registered.
+    /// </summary>
+    /// <param name="services">Services to use when instantiating decompression providers.</param>
+    /// <param name="logger">An instance of <see cref="ILogger"/>.</param>
+    /// <param name="options">The options for this instance.</param>
+    public RequestDecompressionProvider(
+        IServiceProvider services,
+        ILogger<RequestDecompressionProvider> logger,
+        IOptions<RequestDecompressionOptions> options)
+    {
+        if (services == null)
+        {
+            throw new ArgumentNullException(nameof(services));
+        }
+
+        if (logger == null)
+        {
+            throw new ArgumentNullException(nameof(logger));
+        }
+
+        if (options == null)
+        {
+            throw new ArgumentNullException(nameof(options));
+        }
+
+        _logger = logger;
+
+        var requestDecompressionOptions = options.Value;
+
+        _providers = requestDecompressionOptions.Providers.ToArray();
+        if (_providers.Length == 0)
+        {
+            _providers = new IDecompressionProvider[]
+            {
+                new DecompressionProviderFactory(typeof(BrotliDecompressionProvider)),
+                new DecompressionProviderFactory(typeof(DeflateDecompressionProvider)),
+                new DecompressionProviderFactory(typeof(GzipDecompressionProvider))
```

Comment:
Why not create the providers directly in this case?
```suggestion
                new BrotliDecompressionProvider(),
                new DeflateDecompressionProvider(),
                new GzipDecompressionProvider(),
```

Reply:
I've updated this to directly instantiate the default providers.